File: android/app/src/main/java/com/zylod/wholesale/data/api/ServerConfig.kt
package com.zylod.wholesale.data.api

import android.content.Context
import android.util.Log
import com.zylod.wholesale.BuildConfig
import com.zylod.wholesale.session.WebProvenance
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit
import org.json.JSONObject

/**
 * Resolves the active backend from BuildConfig.SERVER_ENDPOINTS (same list the
 * WebView shell probes) and caches it under the same prefs file so both shells
 * agree. Any HTTP response (even 404) proves the host is alive.
 *
 * F1 — deterministic, AUDITABLE endpoint discovery:
 *  - every probe is logged with its HTTP status and whether the responder is
 *    a genuine Zylod server (parses as the /api/app/version payload and
 *    reports a bundle identity);
 *  - the selection reason is logged ("first-alive-in-order" is the frozen
 *    policy — this fix changes VISIBILITY, not selection semantics);
 *  - an endpoint change between resolves is logged prominently — endpoint
 *    switching is never silent (owner hard rule);
 *  - each probe also records the responder's web bundle identity so the
 *    chain "exact build → exact web source → exact web bundle commit" is
 *    answerable from `adb logcat -s ZylodProvenance` alone (docs/PROVENANCE.md).
 */
object ServerConfig {
    private const val PREFS = "zylod_config"
    private const val KEY_BASE = "native_base_url"
    private const val PROBE_BUDGET_MS = 4_000L

    /** Result of probing ONE candidate endpoint. */
    data class ProbeResult(
        val alive: Boolean,
        val httpStatus: Int?,
        /** true when the response parses as a Zylod /api/app/version payload. */
        val zylodPayload: Boolean,
    )

    /** The endpoint's reported bundle identity from the last resolve (diagnostics). */
    @Volatile
    var lastProbeMatrix: Map<String, ProbeResult> = emptyMap()
        private set

    fun cached(context: Context): String? =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getString(KEY_BASE, null)

    suspend fun resolve(context: Context): String = withContext(Dispatchers.IO) {
        val candidates = BuildConfig.SERVER_ENDPOINTS.toList()
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

        coroutineScope {
            val probes = candidates.associateWith { async { probe(it) } }
            val results = probes.mapValues { (_, deferred) ->
                runCatching { deferred.await() }.getOrDefault(ProbeResult(alive = false, httpStatus = null, zylodPayload = false))
            }
            lastProbeMatrix = results

            // Ordered probe matrix — the owner sees exactly why a host won.
            results.forEach { (endpoint, result) ->
                val status = result.httpStatus?.toString() ?: "unreachable"
                Log.i(
                    WebProvenance.LOG_TAG,
                    "probe endpoint=$endpoint status=$status zylodPayload=${result.zylodPayload} alive=${result.alive}",
                )
            }

            val alive = candidates.firstOrNull { results[it]?.alive == true }
            val previous = cached(context)
            val chosen = alive ?: previous ?: candidates.first()
            val reason = when {
                alive != null -> "first-alive-in-order"
                previous != null -> "all-dead-cache-fallback"
                else -> "all-dead-default-first"
            }
            if (previous != null && previous != chosen) {
                Log.w(WebProvenance.LOG_TAG, "ENDPOINT CHANGED $previous -> $chosen (reason=$reason)")
            }
            Log.i(WebProvenance.LOG_TAG, "selected endpoint=$chosen reason=$reason")
            prefs.edit().putString(KEY_BASE, chosen).apply()
            chosen
        }
    }

    private val probeClient = OkHttpClient.Builder()
        .connectTimeout(PROBE_BUDGET_MS, TimeUnit.MILLISECONDS)
        .readTimeout(PROBE_BUDGET_MS, TimeUnit.MILLISECONDS)
        .build()

    /**
     * Aliveness stays "any HTTP response" (frozen selection semantics), but
     * the probe now ALSO fingerprints the responder: a successful
     * /api/app/version body that parses as the Zylod payload proves the host
     * is a genuine Zylod server, not a captive portal or a bare stub.
     */
    private fun probe(endpoint: String): ProbeResult = try {
        probeClient.newCall(
            Request.Builder()
                .url(endpoint.removeSuffix("/") + "/api/app/version")
                .header("User-Agent", "ZylodNative/${BuildConfig.VERSION_NAME}")
                .build()
        ).execute().use { response ->
            var zylodPayload = false
            if (response.isSuccessful) {
                zylodPayload = runCatching {
                    val body = response.peekBody(64_000).string()
                    val json = JSONObject(body)
                    json.optBoolean("success", false) || json.has("bundle")
                }.getOrDefault(false)
            }
            ProbeResult(alive = true, httpStatus = response.code, zylodPayload = zylodPayload)
        }
    } catch (_: Exception) {
        ProbeResult(alive = false, httpStatus = null, zylodPayload = false)
    }
}

File: ios/Zylod/Networking/ServerConfig.swift
import Foundation
import os

/// Resolves the active backend from the same candidate list the Android shell
/// and WebView probe (build-config specific), caching the winner.
///
/// F1 — deterministic, AUDITABLE endpoint discovery (Android parity; owner
/// deployment order: discovery is "acceptable only when the selected server
/// actually serves the matching web bundle"):
///  - every candidate is probed on /api/app/version; the probe records the
///    HTTP status, whether the responder is a GENUINE Zylod server (a 2xx
///    body that reports a bundle identity — docs/PROVENANCE.md §3) and WHICH
///    bundle commit it serves;
///  - SELECTION: the first candidate IN THE FROZEN ORDER that is genuine
///    wins. A host that answers but serves no Zylod bundle identity (redirect
///    stub, captive portal, 404) is skipped and logged — it can no longer
///    beat a genuine server further down the list. Only when NO candidate is
///    genuine does discovery fall back (first-alive → cache → first), with
///    the reason logged and the provenance gate expected to refuse the
///    non-Zylod responder;
///  - an endpoint change between resolves is logged prominently — endpoint
///    switching is never silent (owner hard rule).
/// Owner audit:
///     log stream --predicate 'category == "ZylodProvenance"'
enum ServerConfig {
    static let candidates: [String] = {
        #if DEBUG
        return [
            "http://192.168.43.79:3000",
            "http://10.0.2.2:3000",
            "http://localhost:3000",
            "https://jugular-winnings-backfield.ngrok-free.dev",
            "https://zylod.com",
        ]
        #else
        return ["https://zylod.com"]
        #endif
    }()

    private static let defaults = UserDefaults.standard
    private static let cacheKey = "native_base_url"
    private static let logger = Logger(
        subsystem: WebProvenance.logSubsystem,
        category: WebProvenance.logCategory,
    )

    /// Result of probing ONE candidate endpoint.
    struct ProbeResult {
        var alive: Bool
        var httpStatus: Int?
        /// true when the responder is a genuine Zylod server: 2xx /api/app/version body reporting a bundle identity.
        var zylodPayload: Bool
        /// The bundle commit the endpoint reports (diagnostics; "" when absent).
        var servedCommit: String
    }

    static func cached() -> String? {
        defaults.string(forKey: cacheKey)
    }

    /// Drops the cached winner so the next resolve() re-probes every candidate.
    /// Retry-after-failure must not loop forever on a dead cached host —
    /// parity with android WebScreen retry → ServerConfig.resolve.
    static func invalidateCache() {
        defaults.removeObject(forKey: cacheKey)
    }

    static func resolve() async -> String {
        var results: [Int: ProbeResult] = [:]
        await withTaskGroup(of: (Int, ProbeResult).self) { group in
            for (index, candidate) in candidates.enumerated() {
                group.addTask {
                    let result = await probe(candidate)
                    return (index, result)
                }
            }
            for await (index, result) in group {
                results[index] = result
            }
        }

        // Ordered probe matrix — the owner sees exactly why a host won.
        for (index, candidate) in candidates.enumerated() {
            let result = results[index] ?? ProbeResult(
                alive: false, httpStatus: nil, zylodPayload: false, servedCommit: "")
            let status = result.httpStatus.map(String.init) ?? "unreachable"
            let served = result.servedCommit.isEmpty ? "none" : String(result.servedCommit.prefix(7))
            logger.info("probe endpoint=\(candidate, privacy: .public) status=\(status, privacy: .public) zylodPayload=\(result.zylodPayload, privacy: .public) servedBundle=\(served, privacy: .public) alive=\(result.alive, privacy: .public)")
        }

        // SELECTION (owner deployment order): the first GENUINE Zylod server
        // in the frozen candidate order wins; a bundle-less responder can
        // never beat a genuine server further down the list.
        let genuineIndex = candidates.indices.first { results[$0]?.zylodPayload == true }
        let firstAliveIndex = candidates.indices.first { results[$0]?.alive == true }
        let previous = cached()
        let chosen = genuineIndex.map { candidates[$0] }
            ?? firstAliveIndex.map { candidates[$0] }
            ?? previous
            ?? candidates[0]
        let reason: String
        if genuineIndex != nil {
            reason = "first-genuine-zylod-in-order"
        } else if firstAliveIndex != nil {
            reason = "no-genuine-endpoint-first-alive(gate-will-refuse)"
        } else if previous != nil {
            reason = "all-dead-cache-fallback"
        } else {
            reason = "all-dead-default-first"
        }
        if let previous, previous != chosen {
            logger.warning("ENDPOINT CHANGED \(previous, privacy: .public) -> \(chosen, privacy: .public) (reason=\(reason, privacy: .public))")
        }
        logger.info("selected endpoint=\(chosen, privacy: .public) reason=\(reason, privacy: .public)")
        defaults.set(chosen, forKey: cacheKey)
        return chosen
    }

    /// A candidate is GENUINE when /api/app/version answers 2xx with a JSON
    /// body that reports a non-empty bundle commit (PROVENANCE.md §3 — a
    /// genuine Zylod server always identifies the web bundle it serves).
    private static func probe(_ endpoint: String) async -> ProbeResult {
        guard let url = URL(string: endpoint + "/api/app/version") else {
            return ProbeResult(alive: false, httpStatus: nil, zylodPayload: false, servedCommit: "")
        }
        var request = URLRequest(url: url, timeoutInterval: 6)
        request.setValue("ZylodNative/\(WebProvenance.versionName())", forHTTPHeaderField: "User-Agent")
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            let http = response as? HTTPURLResponse
            var servedCommit = ""
            if let http, (200..<300).contains(http.statusCode) {
                let json = (try? JSONSerialization.jsonObject(with: data)).flatMap { $0 as? [String: Any] }
                servedCommit = ((json?["bundle"] as? [String: Any])?["commit"] as? String ?? "")
                    .trimmingCharacters(in: .whitespacesAndNewlines)
            }
            return ProbeResult(
                alive: true,
                httpStatus: http?.statusCode,
                zylodPayload: !servedCommit.isEmpty,
                servedCommit: servedCommit,
            )
        } catch {
            return ProbeResult(alive: false, httpStatus: nil, zylodPayload: false, servedCommit: "")
        }
    }
}
